util/mock/context.go

```diff
@@ -23,21 +23,24 @@ import (
 	"github.com/pingcap/tidb/kv"
 	"github.com/pingcap/tidb/sessionctx/variable"
 	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/kvcache"
 	goctx "golang.org/x/net/context"
 )
 
 var _ context.Context = (*Context)(nil)
 
 // Context represents mocked context.Context.
 type Context struct {
-	values      map[fmt.Stringer]interface{}
-	txn         kv.Transaction // mock global variable
-	Store       kv.Storage     // mock global variable
-	sessionVars *variable.SessionVars
-	mux         sync.Mutex // fix data race in ddl test.
-	ctx         goctx.Context
-	cancel      goctx.CancelFunc
-	sm          util.SessionManager
+	values       map[fmt.Stringer]interface{}
+	txn          kv.Transaction // mock global variable
+	Store        kv.Storage     // mock global variable
+	sessionVars  *variable.SessionVars
+	mux          sync.Mutex // fix data race in ddl test.
+	ctx          goctx.Context
+	cancel       goctx.CancelFunc
+	sm           util.SessionManager
+	pcache       *kvcache.SimpleLRUCache
```

Comment:
Can we use a global cache shared by every session in a tidb-server for prepare statements ?

Reply:
It might not be a good idea for the performance perspective. I've observed that the lock overhead for the global cache seriously compromises the performance advantages of the plan cache.

Reply:
How about using a sharded lru cache ? If we use a single cache in every session, the memory usage of a tidb server can be very large.

Reply:
What about providing both options?

Reply:
It is possible if there is a sharded lru cache not having any lock overhead.
Is there any reference for the sharded lru cache?

Reply:
It might be a trade-off problem between performance and memory usage. Since "tidb-server" is CPU bound, I think that performance is more important than memory usage. 

BTW, in my test, TiDB monitoring dashboard showed that "tidb-server" only used up to 80 MB including the memory overhead. Of course, it depends on the parameters.